multiaddr/codec.py

```diff
@@ -71,47 +88,27 @@ def bytes_to_string(buf):
     return '/'.join(st)
 
 
-int_to_hex = None
-encode_big_endian_16 = None
-
-
 def address_string_to_bytes(proto, addr_string):
-    global int_to_hex
-    if int_to_hex is None:
-        from .util import int_to_hex
-
-    global encode_big_endian_16
-    if encode_big_endian_16 is None:
-        from .util import encode_big_endian_16
-
     if proto.code == P_IP4:  # ipv4
         try:
-            ip = IPAddress(addr_string)
-            if ip.version != 4:
-                raise ValueError("failed to parse ip4 addr: %s" % addr_string)
-            return int_to_hex(int(ip), 8)
+            return IPAddress(addr_string, version=4).packed
         except Exception:
```

Comment:
Though it is not included in the PR, would you mind changing it to `netaddr.core.AddrFormatError`?
```suggestion
         except netaddr.core.AddrFormatError:
```

Reply:
I'll add this #38 instead to avoid unnecessary merging as I do modify this there.

That said, I think exception handling could use some more love beyond this:

  * Use [`raise from`](https://docs.python.org/3.7/reference/simple_stmts.html#index-29) everywhere were we do this *catch then raise something else* kind of thing to properly document the exception chain involved
  * Add some more specialized exceptions (ParsingError, InternalStateError, …) that indirectly derive from `ValueError` for back-compat
  * Revisit all `except`-clauses to see if their match patterns make sense (as you suggested here)

Does this sound like something you'd want? (Otherwise I'll just include your suggesting in #38.)

Reply:
It's really great idea! IMO it makes great sense to switch to this handling pattern. I will create an issue for it. Thank you to suggest it:)